src/classes/ExtendedGraph.class.php

```diff
@@ -171,24 +177,60 @@ public function add_literal_triple($s, $p, $o, $lang = null, $dt = null) {
      * @return bool
      */
     private function _add_triple($s, $p, Array $o_info) {
-        if (!isset($this->_index[$s])) {
-            $this->_index[$s] = array();
-            $this->_index[$s][$p] = array( $o_info );
-            return true;
-        }
-        elseif (!isset($this->_index[$s][$p])) {
-            $this->_index[$s][$p] = array( $o_info);
-            return true;
-        }
-        else {
-            if ( ! in_array( $o_info, $this->_index[$s][$p] ) ) {
-                $this->_index[$s][$p][] = $o_info;
+        // Make sure the subject is not an empty string
+        if($s === ""){
+            throw new \Tripod\Exceptions\Exception("The subject cannot be an empty string");
+        }
+        // The value $o should already have been validated by this point
+        // It's validation differs depending on whether it is a literal or resource
+        // So just check the subject and predicate here...
+        if($this->isValidTripleValue($s) && $this->isValidTripleValue($p)) {
+            if (!isset($this->_index[$s])) {
+                $this->_index[$s] = array();
+                $this->_index[$s][$p] = array($o_info);
                 return true;
+            } elseif (!isset($this->_index[$s][$p])) {
+                $this->_index[$s][$p] = array($o_info);
+                return true;
+            } else {
+                if (!in_array($o_info, $this->_index[$s][$p])) {
+                    $this->_index[$s][$p][] = $o_info;
+                    return true;
+                }
             }
         }
         return false;
     }
 
+
+    /**
+     * Check if a triple value is valid.
+     *
+     * Ideally a valid literal value should be a string
+     * but accepting scalars so we can handle legacy data
+     * which was not type-checked.
+     *
+     * @return bool
+     */
+    protected function isValidLiteralValue($value){
+        if(!is_scalar($value)){
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check if a triple value is valid.
+     *
+     * @return bool
+     */
+    protected function isValidTripleValue($value){
```

Comment:
Should I be breaking this out into ``isValidSubject``, ``isValidPredicate``, and ``isValidVaue``?  It might make sense since we are now ensuring that the subject is not an empty string, as we can just add the check into there...